dot/state/grandpa_changes.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package state
+
+import (
+	"fmt"
+
+	"github.com/ChainSafe/gossamer/dot/types"
+	"github.com/ChainSafe/gossamer/lib/common"
+)
+
+type isDescendantOfFunc func(parent, child common.Hash) (bool, error)
+
+type pendingChange struct {
+	bestFinalizedNumber uint32
+	delay               uint32
+	nextAuthorities     []types.Authority
+	announcingHeader    *types.Header
+}
+
+func (p pendingChange) String() string {
+	return fmt.Sprintf("announcing header: %s (%d), delay: %d, next authorities: %d",
+		p.announcingHeader.Hash(), p.announcingHeader.Number, p.delay, len(p.nextAuthorities))
+}
+
+func (p *pendingChange) effectiveNumber() uint {
+	return p.announcingHeader.Number + uint(p.delay)
+}
+
+type orderedPendingChanges []*pendingChange
+
+func (o orderedPendingChanges) Len() int      { return len(o) }
+func (o orderedPendingChanges) Swap(i, j int) { o[i], o[j] = o[j], o[i] }
+
+// Less order by effective number and then by block number
+func (o orderedPendingChanges) Less(i, j int) bool {
+	return o[i].effectiveNumber() < o[j].effectiveNumber() &&
+		o[i].announcingHeader.Number < o[j].announcingHeader.Number
+}
+
+type pendingChangeNode struct {
+	change *pendingChange
+	nodes  []*pendingChangeNode
+}
+
+func (c *pendingChangeNode) importScheduledChange(blockHash common.Hash, blockNumber uint, pendingChange *pendingChange,
+	isDescendantOf isDescendantOfFunc) (imported bool, err error) {
+	announcingHash := c.change.announcingHeader.Hash()
+
+	if blockHash.Equal(announcingHash) {
+		return false, errDuplicateHashes
+	}
+
+	if blockNumber <= c.change.announcingHeader.Number {
+		return false, nil
+	}
+
+	for _, childrenNodes := range c.nodes {
```

Comment:
rename from `childrenNodes` to `pendingChangeNode`?

Reply:
I prefree

Reply:
I prefer to keep `childrenNodes` as a reference to inner/nested nodes